llvm/tools/sycl-post-link/sycl-post-link.cpp

```diff
@@ -295,15 +296,95 @@ saveResultModules(std::vector<std::unique_ptr<Module>> &ResModules) {
+  if (FuncName == "__devicelib_assert_fail") {
+    return DeviceLibAssert;
+  }
+  size_t Len =
+      sizeof(CmathDeviceLibFunctions) / sizeof(CmathDeviceLibFunctions[0]);
+  if (std::binary_search(CmathDeviceLibFunctions, CmathDeviceLibFunctions + Len,
+                         FuncName)) {
+    return DeviceLibCmath;
+  }
+  Len =
+      sizeof(Cmath64DeviceLibFunctions) / sizeof(Cmath64DeviceLibFunctions[0]);
+  if (std::binary_search(Cmath64DeviceLibFunctions,
+                         Cmath64DeviceLibFunctions + Len, FuncName)) {
+    return DeviceLibCmath64;
+  }
+  Len =
+      sizeof(ComplexDeviceLibFunctions) / sizeof(ComplexDeviceLibFunctions[0]);
+  if (std::binary_search(ComplexDeviceLibFunctions,
+                         ComplexDeviceLibFunctions + Len, FuncName)) {
+    return DeviceLibComplex;
+  }
+  Len = sizeof(Complex64DeviceLibFunctions) /
+        sizeof(Complex64DeviceLibFunctions[0]);
+  if (std::binary_search(Complex64DeviceLibFunctions,
+                         Complex64DeviceLibFunctions + Len, FuncName)) {
+    return DeviceLibComplex64;
+  }
+  return 0;
+}
+
+// For each device image module, we go through all functions which meets
+// 1. The function name has prefix "__devicelib_"
+// 2. The function has SPIR_FUNC calling convention
+// 3. The function is declaration which means it doesn't have function body
+static uint32_t getModuleReqMask(const std::unique_ptr<Module> &MPtr) {
+  uint32_t ReqMask = 0;
+  uint32_t DeviceLibBits = 0;
+  for (const Function &SF : *MPtr) {
+    if (SF.getName().startswith("__devicelib_") &&
+        (SF.getCallingConv() == CallingConv::SPIR_FUNC) && SF.isDeclaration()) {
+      DeviceLibBits = getDeviceLibBits(SF.getName().str());
+      ReqMask |= DeviceLibBits;
+    }
+  }
+  return ReqMask;
+}
+
+static void
+getDeviceLibReqMasks(const std::vector<std::unique_ptr<Module>> &ResModules,
+                     std::vector<uint32_t> &DeviceLibReqMaskVec) {
+  for (auto &MPtr : ResModules) {
+    uint32_t ModuleReqMask = getModuleReqMask(MPtr);
+    DeviceLibReqMaskVec.push_back(ModuleReqMask);
+  }
+}
+
 static string_vector
-saveSpecConstantIDMaps(const std::vector<SpecIDMapTy> &Maps) {
+saveDeviceImageProperty(const std::vector<uint32_t> ReqMaskVec,
+                        const std::vector<SpecIDMapTy> &Maps) {
   string_vector Res;
-
-  for (size_t I = 0; I < Maps.size(); ++I) {
+  bool saveSpecIDMaps =
+      (Maps.size() != 0) && (ReqMaskVec.size() == Maps.size());
+  for (size_t I = 0; I < ReqMaskVec.size(); ++I) {
     std::string SCFile = makeResultFileName(".prop", I);
     llvm::util::PropertySetRegistry PropSet;
-    PropSet.add(llvm::util::PropertySetRegistry::SYCL_SPECIALIZATION_CONSTANTS,
-                Maps[I]);
+    std::map<StringRef, uint32_t> reqMaskEntry;
+    reqMaskEntry["devicelib_req_mask"] = ReqMaskVec[I];
+    PropSet.add(llvm::util::PropertySetRegistry::SYCL_DEVICELIB_REQ_MASK,
+                reqMaskEntry);
```

Comment:
This always adds a map with a single entry?

Reply:
Yes, for each sub-module, we will add one mask.